### PR TITLE
[WIP]Flax training script for controlnet

### PR DESCRIPTION
**Pull request by yiyixuxu**
wandb test run here: https://wandb.ai/yiyixu/train_controlnet_flax/runs/jye8xcs9?workspace=user-yiyixu
```
wget https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/controlnet_training/conditioning_image_1.png
```

```bash
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="control_out"

python3 train_controlnet_flax.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \
 --validation_image "./conditioning_image_1.png"  \
 --validation_prompt "red circle with blue background" \
 --validation_steps=1000 \
 --train_batch_size=1 \
 --revision="flax"\
 --report_to="wandb"
```

### CODE REVIEW

**Comment by jingyangcarl** [comment]
Hi, awesome works!! I run the code and find the following @yiyixuxu, any chance you can share some ideas on this? #2822

File: examples/controlnet/train_controlnet_flax.py [hunk 1]
```diff
@@ -0,0 +1,945 @@
...
+    )
+
+    optimizer = optax.chain(
+        optax.clip_by_global_norm(args.max_grad_norm),
+        adamw,
+    )
+
+    state = train_state.TrainState.create(apply_fn=controlnet.__call__, params=controlnet_params, tx=optimizer)
+
+    noise_scheduler = FlaxDDPMScheduler(
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply examples/controlnet/train_controlnet_flax.py:711]: this should be saver than customizing everything

File: examples/controlnet/train_controlnet_flax.py [hunk 2]
```diff
@@ -0,0 +1,945 @@
...
+    noise_scheduler_state = noise_scheduler.create_state()
+
+    # Initialize our training
+    validation_rng, train_rngs = jax.random.split(rng)
+    train_rngs = jax.random.split(train_rngs, jax.local_device_count())
+
+    def train_step(state, unet_params, text_encoder_params, vae_params, batch, train_rng):
+        # reshape batch, add grad_step_dim if gradient_accumulation_steps > 1
+        if args.gradient_accumulation_steps > 1:
+            G = args.gradient_accumulation_steps
```
Comment (patrickvonplaten) [inline examples/controlnet/train_controlnet_flax.py:723]:
```suggestion
            grad_steps = args.gradient_accumulation_steps
```

File: examples/controlnet/train_controlnet_flax.py [hunk 3]
```diff
@@ -0,0 +1,945 @@
...
+            batch = jax.tree_map(lambda x: x.reshape((G, x.shape[0] // G) + x.shape[1:]), batch)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply examples/controlnet/train_controlnet_flax.py:724]: Not a fan of single capital letters

**Review by patrickvonplaten — APPROVED** [review]
Super nice! Feel free to merge :-)

**Comment by patrickvonplaten** [comment]
BTW, could you also add this:

```
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="control_out"

python3 train_controlnet_flax.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \
 --validation_image "./conditioning_image_1.png"  \
 --validation_prompt "red circle with blue background" \
 --validation_steps=12500 \
 --train_batch_size=1 \
 --revision="flax"\
 --report_to="wandb"
```

to the README here: https://github.com/huggingface/diffusers/blob/d9cf213e9b8cc59be953e1da7df916346612aef3/examples/controlnet/README.md

File: examples/controlnet/train_controlnet_flax.py [hunk 4]
```diff
@@ -0,0 +1,945 @@
...
+    else:
+        raise NotImplementedError("No tokenizer specified!")
+
+    # Get the datasets: you can either provide your own training and evaluation files (see below)
+    train_dataset = make_train_dataset(args, tokenizer)
+    total_train_batch_size = args.train_batch_size * jax.local_device_count() * args.gradient_accumulation_steps
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset,
+        shuffle=True,
```
Comment (sayakpaul) [inline examples/controlnet/train_controlnet_flax.py:620]:
```suggestion
        shuffle=not args.streaming,
```

**Comment by yiyixuxu** [comment]
I've been running into this issue this afternoon https://github.com/huggingface/diffusers/issues/2851
but for the last 2 times, I was able to save the model to the hub (despite the error message) - I will leave that issue open for now in case we run into this issue in the future 

see an example repo here with an automatically generated model card here (checkpoint with 4k steps) 
https://huggingface.co/YiYiXu/fill-circle-controlnet

I will merge this in now so that more people can start to test it. Will open a separate PR to add intermediate checkpointing and make other improvements based on feedback.

**Comment by sayakpaul** [comment]
Thanks, @yiyixuxu. Do you plan on integrating the streaming support as we were discussing in https://github.com/huggingface/diffusers/issues/2840?

**Comment by yiyixuxu** [comment]
Thanks @sayakpaul  
I'm taking a look right now, but you're much more of an expert in this than me:)  I think it's up to you and @patrickvonplaten

**Comment by sayakpaul** [comment]
> I'm taking a look right now, but you're much more of an expert in this than me:)

Actually, I don't believe the changes for enabling streaming will be long. I already mentioned a couple of pointers in https://github.com/huggingface/diffusers/issues/2840#issue-1641759298. Maybe you could try following the changes I made in [here](https://gist.github.com/sayakpaul/b7ba991bf15e302ad0c9bbfe6e1e7f83#file-train_dummy-py) in your next PR? Happy to guide you if need be.

**Comment by yiyixuxu** [comment]
@sayakpaul yeah a PR please! happy to test it out :)

**Comment by sayakpaul** [comment]
Perf! Will have it opened as soon as possible.

**Pull request merged**